tcpdevmem.cu

```diff
@@ -0,0 +1,754 @@
+};
+
+#define DMA_BUF_CREATE_PAGES \
+  _IOW(DMA_BUF_BASE, 2, struct dma_buf_create_pages_info)
+
+#define DMA_BUF_PAGES_BIND_RX \
+  _IOW(DMA_BUF_BASE, 3, struct dma_buf_pages_bind_rx_queue)
+
+// devmemvec represents a fragment of payload that is received on the socket.
+struct devmemvec {
+  // frag_offset is the offset in the registered memory.
+  __u32 frag_offset;
+  // frag size is the size of the payload.
+  __u32 frag_size;
+  // frag_token is an identifier for this fragment and it can be used to return
+  // the memory back to kernel.
+  __u32 frag_token;
+};
+
+// devmemtoken represents a range of tokens. It is used to return the fragment
+// memory back to the kernel.
+struct devmemtoken {
+  __u32 token_start;
+  __u32 token_count;
+};
+
+struct TcpdRxBlock {
+  uint64_t gpu_offset;
+  size_t size;
+  uint64_t paddr;
+};
+
+struct udmabuf_create {
+  uint32_t memfd;
+  uint32_t flags;
+  uint64_t offset;
+  uint64_t size;
+};
+#define UDMABUF_CREATE _IOW('u', 0x42, struct udmabuf_create)
+
+int memfd_create(const char *name, unsigned int flags)
+{
+	return syscall(__NR_memfd_create, name, flags);
+}
+
+/* Fills buf of size n with a repeating sequence of 1 to 111 inclusive
+ */
+void fill_tx_buffer(void *buf, size_t n) {
+#define BUFSIZE 3996
+  unsigned char src_buf[BUFSIZE];
+  int ptr = 0, i = 0;
+
+  while (i < BUFSIZE) {
+    src_buf[i] = (i % LAST_PRIME) + 1;
+    i++;
+  }
+
+  while (ptr*BUFSIZE + BUFSIZE < n) {
+    cudaMemcpy((char *)buf + ptr*BUFSIZE, &src_buf, BUFSIZE, cudaMemcpyHostToDevice);
+    ptr++;
+  }
+
+  i = ptr*BUFSIZE;
+  while (i < n) {
+    cudaMemset((char *)buf + i, (i % LAST_PRIME) + 1, 1);
+    i++;
+  }
+}
+
+__global__ void scatter_copy_kernel(long3* scatter_list, uint8_t* dst,
+                                    uint8_t* src) {
+  int block_idx = blockIdx.x;
+  long3 blk = scatter_list[block_idx];
+  long dst_off = blk.x;
+  long src_off = blk.y;
+  long sz = blk.z;
+
+  int thread_sz = sz / blockDim.x;
+  int rem = sz % blockDim.x;
+  bool extra = (threadIdx.x < rem);
+  int thread_offset = sz / blockDim.x * threadIdx.x;
+  thread_offset += (extra) ? threadIdx.x : rem;
+
+  for (int i = 0; i < thread_sz; i++) {
+    dst[dst_off + thread_offset + i] = src[src_off + thread_offset + i];
+  }
+  if (extra) {
+    dst[dst_off + thread_offset + thread_sz] =
+        src[src_off + thread_offset + thread_sz];
+  }
+}
+
+void gather_rx_data(struct tcpdevmem_cuda_mbuf *tmbuf) {
+  int ret;
+  void *gpu_scatter_list_ = tmbuf->gpu_scatter_list_;
+  std::vector<long3> *scattered_data_ = (std::vector<long3> *)tmbuf->scattered_data_;
+  void *gpu_rx_mem_ = tmbuf->gpu_rx_mem_;
+  void *rx_buff_ = tmbuf->gpu_gen_mem_;
+
+  ret = cudaMemcpyAsync(gpu_scatter_list_,
+                        scattered_data_->data(),
+                        scattered_data_->size() * sizeof(long3),
+                        cudaMemcpyHostToDevice);
+  if (ret)
+    return;
+
+  scatter_copy_kernel<<<scattered_data_->size(), 256, 0>>>(
```

Comment:
consider using `cudaGetLastError()` to confirm if the kernel successfully launched here.

Reply:
Also consider replacing "256" with a macro